### ISSUE TITLE
[FELIX] - Tagging loss becomes NaN

### ISSUE THREAD

**Issue by m-wolfsteiner** [google-research/google-research#1635]
Hello,

when I run the DiscoFuse running example for FELIX, during the training of the tagging model, the loss becomes NaN at some point. This doesn't happen for the insertion model. 

I am using the tensorflow version 2.9.1 (I installed the requirements.txt with pip).

This is my call of the tagging model:
```
python3 run_felix.py \
    --train_file="${OUTPUT_DIR}/train.tfrecord" \
    --eval_file="${OUTPUT_DIR}/dev.tfrecord" \
    --model_dir_tagging="${OUTPUT_DIR}/model_tagging" \
    --bert_config_tagging="${DISCOFUSE_DIR}/felix_config.json" \
    --max_seq_length=128 \
    --num_train_epochs=500 \
    --num_train_examples=450 \
    --num_eval_examples=45 \
    --train_batch_size="16" \
    --eval_batch_size="16" \
    --log_steps="100" \
    --steps_per_loop="100" \
    --train_insertion="False" \
    --use_pointing="${USE_POINTING}" \
    --init_checkpoint="${BERT_BASE_DIR}/bert_model.ckpt" \
    --learning_rate="0.00003" \
    --pointing_weight="1" \
    --use_weighted_labels="True"
```

This is the initial output I am getting from the tagging model:
``` 
Training tagging model
W0702 01:54:05.900973 22854577735488 bert_encoder.py:378] `sequence_length` is a deprecated argument to `BertEncoder`, which has no effect for a while. Please remove `sequence_length` argument.
2023-07-02 01:54:05.908122: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-07-02 01:54:06.609735: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1532] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 10415 MB memory:  -> device: 0, name: NVIDIA GeForce GTX 1080 Ti, pci bus id: 0000:af:00.0, compute capability: 6.1
I0702 01:54:10.107235 22854577735488 optimization.py:87] using Adamw optimizer
I0702 01:54:10.107491 22854577735488 optimization.py:140] gradient_clip_norm=1.000000
WARNING:tensorflow:The argument `steps_per_execution` is no longer experimental. Pass `steps_per_execution` instead of `experimental_steps_per_execution`.
W0702 01:54:10.107635 22854577735488 training.py:627] The argument `steps_per_execution` is no longer experimental. Pass `steps_per_execution` instead of `experimental_steps_per_execution`.
I0702 01:54:10.427585 22854577735488 run_felix.py:166] Initializing from a BERT checkpoint...
I0702 01:54:52.159135 22854577735488 run_felix.py:183] Starting training from iteration: 0.
WARNING:tensorflow:Model failed to serialize as JSON. Ignoring... 'FelixTagLoss' object has no attribute '_config'
W0702 01:54:52.172672 22854577735488 callbacks.py:2061] Model failed to serialize as JSON. Ignoring... 'FelixTagLoss' object has no attribute '_config'
Epoch 1/500
WARNING:tensorflow:Gradients do not exist for variables ['pooler_transform/kernel:0', 'pooler_transform/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss`argument?
W0702 01:54:54.478482 22854577735488 utils.py:76] Gradients do not exist for variables ['pooler_transform/kernel:0', 'pooler_transform/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss`argument?
WARNING:tensorflow:Gradients do not exist for variables ['pooler_transform/kernel:0', 'pooler_transform/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss`argument?
W0702 01:54:57.368436 22854577735488 utils.py:76] Gradients do not exist for variables ['pooler_transform/kernel:0', 'pooler_transform/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss`argument?
WARNING:tensorflow:Gradients do not exist for variables ['pooler_transform/kernel:0', 'pooler_transform/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss`argument?
W0702 01:55:00.801969 22854577735488 utils.py:76] Gradients do not exist for variables ['pooler_transform/kernel:0', 'pooler_transform/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss`argument?
WARNING:tensorflow:Gradients do not exist for variables ['pooler_transform/kernel:0', 'pooler_transform/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss`argument?
W0702 01:55:03.620443 22854577735488 utils.py:76] Gradients do not exist for variables ['pooler_transform/kernel:0', 'pooler_transform/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss`argument?
2023-07-02 01:55:19.127899: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:390] Filling up shuffle buffer (this may take a while): 30152 of 50000
2023-07-02 01:55:22.578884: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:415] Shuffle buffer filled.

28/28 [==============================] - ETA: 0s - loss: 11.0296 - tag_loss: 2.6948 - tag_accuracy: 0.1018 - tag_exact_match: 0.0000e+00 - point_loss: 8.3348 - point_accuracy: 0.0254 - point_exact_match: 0.0000e+00 - exact_match: 0.0000e+00 - total_loss: 11.0296
28/28 [==============================] - 67s 2s/step - loss: 11.0296 - tag_loss: 2.6948 - tag_accuracy: 0.1018 - tag_exact_match: 0.0000e+00 - point_loss: 8.3348 - point_accuracy: 0.0254 - point_exact_match: 0.0000e+00 - exact_match: 0.0000e+00 - total_loss: 11.0296 - val_loss: 10.6282 - val_tag_loss: 2.6900 - val_tag_accuracy: 0.0670 - val_tag_exact_match: 0.0000e+00 - val_point_loss: 7.9382 - val_point_accuracy: 0.0280 - val_point_exact_match: 0.0000e+00 - val_exact_match: 0.0000e+00 - val_total_loss: 10.6282
Epoch 2/500

28/28 [==============================] - ETA: 0s - loss: 10.9989 - tag_loss: 2.7371 - tag_accuracy: 0.0885 - tag_exact_match: 0.0000e+00 - point_loss: 8.2618 - point_accuracy: 0.0259 - point_exact_match: 0.0000e+00 - exact_match: 0.0000e+00 - total_loss: 10.9989
28/28 [==============================] - 97s 3s/step - loss: 10.9989 - tag_loss: 2.7371 - tag_accuracy: 0.0885 - tag_exact_match: 0.0000e+00 - point_loss: 8.2618 - point_accuracy: 0.0259 - point_exact_match: 0.0000e+00 - exact_match: 0.0000e+00 - total_loss: 10.9989 - val_loss: 10.5075 - val_tag_loss: 2.6793 - val_tag_accuracy: 0.0682 - val_tag_exact_match: 0.0000e+00 - val_point_loss: 7.8282 - val_point_accuracy: 0.0271 - val_point_exact_match: 0.0000e+00 - val_exact_match: 0.0000e+00 - val_total_loss: 10.5075
Epoch 3/500 
```

After a varying number of epochs, the loss turns to NaN:
```
28/28 [==============================] - ETA: 0s - loss: 0.6629 - tag_loss: 0.4535 - tag_accuracy: 0.8362 - tag_exact_match: 0.0826 - point_loss: 0.2095 - point_accuracy: 0.9404 - point_exact_match: 0.3438 - exact_match: 0.0603 - total_loss: 0.6629
28/28 [==============================] - 8s 300ms/step - loss: 0.6629 - tag_loss: 0.4535 - tag_accuracy: 0.8362 - tag_exact_match: 0.0826 - point_loss: 0.2095 - point_accuracy: 0.9404 - point_exact_match: 0.3438 - exact_match: 0.0603 - total_loss: 0.6629 - val_loss: 0.6517 - val_tag_loss: 0.4527 - val_tag_accuracy: 0.8450 - val_tag_exact_match: 0.0938 - val_point_loss: 0.1990 - val_point_accuracy: 0.9440 - val_point_exact_match: 0.2500 - val_exact_match: 0.0625 - val_total_loss: 0.6517
Epoch 129/500

28/28 [==============================] - ETA: 0s - loss: nan - tag_loss: nan - tag_accuracy: 0.1753 - tag_exact_match: 0.0067 - point_loss: nan - point_accuracy: 0.2869 - point_exact_match: 0.0670 - exact_match: 0.0045 - total_loss: nan
28/28 [==============================] - 8s 296ms/step - loss: nan - tag_loss: nan - tag_accuracy: 0.1753 - tag_exact_match: 0.0067 - point_loss: nan - point_accuracy: 0.2869 - point_exact_match: 0.0670 - exact_match: 0.0045 - total_loss: nan - val_loss: nan - val_tag_loss: nan - val_tag_accuracy: 0.0000e+00 - val_tag_exact_match: 0.0000e+00 - val_point_loss: nan - val_point_accuracy: 0.1120 - val_point_exact_match: 0.0000e+00 - val_exact_match: 0.0000e+00 - val_total_loss: nan
Epoch 130/500

28/28 [==============================] - ETA: 0s - loss: nan - tag_loss: nan - tag_accuracy: 0.0000e+00 - tag_exact_match: 0.0000e+00 - point_loss: nan - point_accuracy: 0.1101 - point_exact_match: 0.0000e+00 - exact_match: 0.0000e+00 - total_loss: nan
28/28 [==============================] - 8s 296ms/step - loss: nan - tag_loss: nan - tag_accuracy: 0.0000e+00 - tag_exact_match: 0.0000e+00 - point_loss: nan - point_accuracy: 0.1101 - point_exact_match: 0.0000e+00 - exact_match: 0.0000e+00 - total_loss: nan - val_loss: nan - val_tag_loss: nan - val_tag_accuracy: 0.0000e+00 - val_tag_exact_match: 0.0000e+00 - val_point_loss: nan - val_point_accuracy: 0.1120 - val_point_exact_match: 0.0000e+00 - val_exact_match: 0.0000e+00 - val_total_loss: nan
Epoch 131/500
```
I have attached the complete output file: [disco_output.txt](https://github.com/google-research/google-research/files/11931475/disco_output.txt)

For the rest of the epochs, only the point_accuracy and the val_point_accuracy are number values. Predictions are generated but it seems like their quality is affected by the NaN. 

I am aware that a NaN loss can have many reasons, so I have tried these strategies with no success:
- Lowering the learning rate: I lowered it to 0.000003, this postponed the point at which the tagging loss becomes NaN, but it still happened
- Clipping the gradient: As I understand it, the gradient is already being clipped, as one of the outputs during training is `using Adamw optimizer` and `gradient_clip_norm=1.000000`
- Batch size: I have made the number of training and eval examples divisible by the batch size, so that the last batch doesn't have a different size. I have also tried the batch sizes 16, 32, 64 and 128.
- I have tried running FELIX on Google Cloud TPUs
- Validity of inputs: I read somewhere that tfrecord doesn't accept NaN values, so they shouldn't occur. But the data could still be the problem.
- I have tried adding jitter to the tag_logits in line 277 of felix_models.py (+ 0.00001)
- I have used varying dataset sizes and epoch numbers

I would be very grateful for any tips or suspicions what the problem is here or to hear how someone else solved this problem! 

Please let me know if I should provide any other information.

**Comment by JaxLee1027** [comment]
Hey, have you solved this problem? I have met the same problem as you.

**Comment by m-wolfsteiner** [comment]
Hi, I am relatively confident that the problem lies in the data and data processing. I noticed that the number of converted insertion examples and converted tagging examples was sometimes not the same. To fix this, I changed line 112 of preprocess_main.py  to ```if example is not None and insertion_example is not None:``` (and deleted line 115). This way, only data points are created, where both insertion and tagging example can be processed. But the NaN still occurred with my own data, and it turned out that some rare data points caused this; I had to manually identify these sentences using binary search on the data set. I did not find out what common factor could have caused this, but it appeared that some input sentences caused the NaN loss. I hope this helps a least a bit, good luck!
